### ISSUE TITLE
gradual increase in memory usage while training RNN with GPU

### ISSUE THREAD

**Issue by apulver** [Theano/Theano#4101]
Using this module:

https://github.com/JonathanRaiman/theano_lstm

and scan to train an RNN.

The memory usage is stable when device=cpu, but it grows slowly when training on a GPU.

**Comment by nouiz** [comment]
This is hard to investigate. The way to start is confirm if this come from
inside a Theano function or not. Then if it is, try to identify which op
cause this. To help investigate this, here is a few questions and think to
try:

Which memory is growing? GPU or CPU? How to do track it?

Can you call "import gc; gc.collect(); gc.collect(); gc.collect()" after
each minibatch and see it this make a difference?

Do you use cudnn? If so, can you try to disable is. dnn.enabled=False

Be sure to use a Theano development version that is up to date. Just in
case it is already fixed:

http://www.deeplearning.net/software/theano/install.html#bleeding-edge-install-instructions

**Comment by apulver** [comment]
It's the CPU memory. Calling gc.collect does not make a difference. I'm using htop to track the memory usage. Using optmizer_excluding = cudnn did not help either. 

Here's some test code. Can you reproduce the error?

`import numpy as np
import theano.tensor as T
import theano
import theano_lstm as lstm
import gc

def softmax(x):
    return T.nnet.softmax(x)[0]

SIZE = 10

INPUT_SIZE = SIZE
LAYERS = [200,200,200]
OUTPUT_SIZE = SIZE

N = 100

network = lstm.StackedCells(INPUT_SIZE, layers=LAYERS, celltype=lstm.RNN, activation=T.tanh)
network.layers.append(lstm.Layer(LAYERS[-1], OUTPUT_SIZE, softmax))

def time_step(x, *prev):
    states = network.forward(x, prev_hiddens=prev) 
    return states

input_sequence = T.fmatrix('input_sequence')
target_sequence = T.ivector('target_sequence') 


results, updates = theano.scan(fn=time_step,
                               outputs_info = [dict(initial=layer.initial_hidden_state, taps=[-1]) for layer in network.layers if hasattr(layer, 'initial_hidden_state')] + [None], 
                               sequences=input_sequence)

reg_cost = T.sum([nparam.norm(L=2) for nparam in network.params])


R_ETA = .001

cost = -T.sum(T.log(results[-1])[T.arange(target_sequence.shape[0]),target_sequence]) + R_ETA*reg_cost

updates, _, _, _, _ = lstm.create_optimization_updates(cost, network.params, rho=0.99, method='sgd')


train_network = theano.function(inputs=[input_sequence, target_sequence],
                                outputs=cost,
                                updates=updates,
                                allow_input_downcast=True)

in1 = np.random.randn(100,10).astype(np.float32)
tg1 = np.random.randint(0, SIZE, size=100).astype(np.int32)
for i in xrange(0, 10000):

    t = train_network(in1,tg1)
    gc.collect()
    print t`

**Comment by nouiz** [comment]
In htop which columns to you watch? RES or SHR?

I don't have time to try to reproduce it now, I'll try tonigh or tomorrow.

**Comment by apulver** [comment]
RES

**Comment by nouiz** [comment]
I'm able to reproduce it.

**Comment by nouiz** [comment]
@abergeron can you look at it?

**Comment by lamblin** [comment]
After investigation, that bug has been present since at least the 0.6 release, and we did not manage to pinpoint which Op in particular, or which other feature, could be responsible for it.
However, we were able to assess that:
- The leak is not present when using only the CPU
- The leak is not present when using the new GPU back-end (the one in `libgpuarray`)

Since we are trying to switch our effort towards the new back-end, rather than supporting the old one, it is unlikely that we will be able to devote more time investigating.

If you have time and are willing to go on investigating, you could start disabling GPU optimizations (the ones moving CPU Ops to the GPU) one by one, and see if you can pinpoint which Op is responsible for the leak.

**Comment by abergeron** [comment]
Just to note that I did that (disabling optimizations) and it lead me to conclude that the problem is either in HostFromGpu, GpuFromHost or the base CudaNdarray code.  However, I don't know more than that.

**Comment by nouiz** [comment]
You can also use the new GPU back end that don't have this problem.

**Comment by apulver** [comment]
Thanks for looking into it. How would one use the new backend?

In the meantime, I've found that using the CPU with a recent openblas version is just as fast, at least for a batch size of 1.

**Comment by nouiz** [comment]
I'll close this issue as you have a good work around. Here is the doc on how to use the new back-end:

http://deeplearning.net/software/theano/tutorial/using_gpu.html#gpuarray-backend
